File: src/cs/PhantasyIslandPythonRemoteControl/radio/ApiModule.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;

namespace PhantasyIslandPythonRemoteControl.Radio
{
    public abstract class ApiModule
    {
        protected readonly RadioManager _rm;
        private string _nowMode = "sync";

        protected ApiModule(RadioManager rm)
        {
            _rm = rm;
        }

        public ApiModule Mode(string mode)
        {
            if (mode == "sync" || mode == "async" || mode == "token")
            {
                _nowMode = mode;
            }
            else
            {
                throw new ArgumentException($"Invalid mode: {mode}");
            }

            return this;
        }

        public string GetNowMode() => _nowMode;

        protected void _Send(string cmd, Dictionary<string, object> data = null)
        {
            _rm.InternalSend(cmd, data);
        }

        protected object? SendInternal(string cmd, Dictionary<string, object>? data = null,
            string? waitCmd = null, double timeout = 3.0,
            Func<JsonElement, object>? postProcessor = null)
        {
            switch (_nowMode)
            {
                case "sync":
                    return _rm.SendAndWaitSync(cmd, data, waitCmd, timeout, postProcessor);
                case "async":
                    return _rm.SendAndWaitAsync(cmd, data, waitCmd, timeout, postProcessor);
                case "token":
                    return _rm.SendWithToken(cmd, data, waitCmd, postProcessor);
                default:
                    throw new InvalidOperationException("Invalid mode");
            }
        }
    }
}

File: src/fs/ApiModule.fs
namespace PhantasyIslandRemoteControl.Radio

open System
open System.Collections.Generic
open System.Text.Json
open System.Threading.Tasks

type IRadioManager =
    abstract member InternalSend: cmd: string * ?data: Dictionary<string, obj> -> unit
    abstract member SendWithToken: cmd: string * ?data: Dictionary<string, obj> * ?waitCmd: string * ?postProcessor: Func<JsonElement, obj> -> WaitToken
    abstract member SendAndWaitSync: cmd: string * ?data: Dictionary<string, obj> * ?waitCmd: string * ?timeout: double * ?postProcessor: Func<JsonElement, obj> -> obj
    abstract member SendAndWaitAsync: cmd: string * ?data: Dictionary<string, obj> * ?waitCmd: string * ?timeout: double * ?postProcessor: Func<JsonElement, obj> -> Async<obj>

[<AbstractClass>]
type ApiModule(rm: IRadioManager) =
    let mutable _nowMode = "sync"
    
    member this.RM = rm

    member this.Mode(mode: string) =
        if mode = "sync" || mode = "async" || mode = "token" then
            _nowMode <- mode
            this
        else
            raise (ArgumentException(sprintf "Invalid mode: %s" mode))

    member this.GetNowMode() = _nowMode

    member this._Send(cmd: string, ?data: Dictionary<string, obj>) =
        rm.InternalSend(cmd, ?data = data)

    member this.SendInternal(cmd: string, ?data: Dictionary<string, obj>, ?waitCmd: string, ?timeout: double, ?postProcessor: Func<JsonElement, obj>) =
        match _nowMode with
        | "sync" -> rm.SendAndWaitSync(cmd, ?data = data, ?waitCmd = waitCmd, ?timeout = timeout, ?postProcessor = postProcessor)
        | "async" -> rm.SendAndWaitAsync(cmd, ?data = data, ?waitCmd = waitCmd, ?timeout = timeout, ?postProcessor = postProcessor) :> obj
        | "token" -> rm.SendWithToken(cmd, ?data = data, ?waitCmd = waitCmd, ?postProcessor = postProcessor) :> obj
        | _ -> raise (InvalidOperationException("Invalid mode"))
